### PR TITLE
Submit moving funds proof

### PR DESCRIPTION
**Pull request by lukasz-zimnoch**
Refs: #179 

This pull request adds a mechanism that can be used to prove a moving funds transaction performed by the wallet on the Bitcoin chain. All logic is written in accordance with the [RFC-1](https://github.com/keep-network/tbtc-v2/blob/main/docs/rfc/rfc-1.adoc#26-wallet-lifecycle) and with respect to the latest requirements updates (specifically, removing commitment challenges).

### Scope of this pull request

The moving funds process consists of several separate steps:
- Step 1: Wallet is put to the `MovingFunds` state by an event
- Step 2: Wallet submits its target wallet commitment for the given moving funds process or directly closes if their balance is below the dust threshold
- Step 3 (happy path): Wallet submits the proof of moving funds
- Step 3 (non-happy path): Wallet fails to submit the proof and a moving funds timeout is reported by a third party 

As mentioned in the introduction, this pull request addresses the moving funds proof (Step 3 - happy path) by adding the `submitMovingFundsProof` function. Step 1 is already implemented and outstanding steps will be addressed in the follow-up pull requests.

Apart from the moving funds proof, this PR also handles some minor improvements and refactoring that touches the moving funds feature more or less. Those are:
- Replace `hash160` with `hash160View`: https://github.com/keep-network/tbtc-v2/pull/184/commits/d5cf9f661004798be4b5c9489542ded6eb688d13
- Rename `validateWalletOutboundTxInput` function as it's no longer just about validation: https://github.com/keep-network/tbtc-v2/pull/184/commits/b08aec0fe48fdea7f8a389964a7956f6969d612f
- Refactor `terminateWallets` to improve the `Wallets` library structure consistency: https://github.com/keep-network/tbtc-v2/pull/184/commits/a7904ff13a9f23d9957cfae8d86092a8e9e5ee1a
- Extract `bridgeFixture` as the same fixture was copy-pasted across Bridge test suites: https://github.com/keep-network/tbtc-v2/pull/184/commits/a57ece939269376ae01213bffa1412a16346128d

### The `submitMovingFundsProof` function

The `submitMovingFundsProof` function can be called only when the target wallets commitment is submitted by a moving funds wallet. Regarding its logic, `submitMovingFundsProof` is very similar to the existing `submitRedemptionProof` function.  It validates the Bitcoin transaction input vector according to the same rules, i.e. it checks whether the single input is actually the current wallet's main UTXO. Regarding the output vector, it expects that the transacted amount is distributed evenly (if the amount is indivisible, there is a tolerance for the remainder) over the target wallets submitted in the commitment. Given the fact that the commitment is stored as a hash (most effective gas-wise), there is also a requirement that the outputs refer to the target wallets in the same order as the wallets were submitted within the commitment. To prevent burning too much of the transaction amount for network fees, the Bitcoin transaction total fee must not exceed the governable `movingFundsTxMaxTotalFee` value. If the validation succeeds, the moving funds proof is accepted and some state changes occur. Specifically, the main UTXO of the source wallet is marked as correctly spent (that prevents fraud accusations) and the source wallet is closed.

### Gas costs

The `submitMovingFundsProof` function takes `142338` on average (min `129200`, max `149540`). However, the ultimate cost strongly depends on the number of used target wallets as it incurs more processing on the output vector side. Presented gas costs are for 1-3 target wallets.

### CODE REVIEW

File: solidity/contracts/bridge/Wallets.sol [hunk 1]
```diff
@@ -517,4 +531,66 @@ library Wallets {
...
+        closeWallet(self, walletPubKeyHash);
+    }
+
+    // TODO: Function for committing target wallets for moving funds.
+    //       Can be called only when `movingFundsRequestedAt` field is
+    //       grater than zero, `movingFundsCommitmentSubmittedAt` and
+    //       `movingFundsTargetWalletsCommitmentHash` fields are zeroed, and
+    //       the caller is a wallet's operator. If validation passes, it must
+    //       set the `movingFundsCommitmentSubmittedAt` and
+    //       `movingFundsTargetWalletsCommitmentHash` fields accordingly.
```
Comment (michalinacienciala) [inline solidity/contracts/bridge/Wallets.sol:583]: Shouldn't we also check if target wallets are LIVE?

Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/Wallets.sol:583]: Not `Live` but `MovingFunds`. However, this is implied by the wallet lifecycle. If `movingFundsRequestedAt` is `> 0` that means something caused the `moveFunds` internal function to be called. That function can be called only on `Live` wallets and makes the state transition to be `MovingFunds`. That said, there is no point to check the wallet state explicitly within the function specified by this TODO as this is redundant in that case.

File: solidity/contracts/bridge/Wallets.sol [hunk 2]
```diff
@@ -517,4 +531,66 @@ library Wallets {
...
+
+    // TODO: Function for challenging target wallets commitments.
+    //       Can only be called if `block.timestamp` is lesser or equal to
+    //       the sum of `movingFundsCommitmentSubmittedAt` field and
+    //       `movingFundsCommitmentChallengePeriod` governance parameter.
+    //       If the challenge is successful, it must reset the
+    //       `movingFundsCommitmentSubmittedAt` and
+    //       `movingFundsTargetWalletsCommitmentHash` fields to their zero
+    //       value and set `movingFundsRequestedAt` to `block.timestamp`
+    //       to adjust the timeout.
```
Comment (michalinacienciala) [inline solidity/contracts/bridge/Wallets.sol:593]: I know the details are not in the scope of this PR, but maybe we could already mention here what will be checked when validating? I.e.:
- are there "closer" LIVE wallets created before target wallets commitment
- is `wallet_balance` / `no_of_selected_wallets` bigger than `wallet_transfer_max` and, if yes, are there any LIVE wallets created before target wallets commitment was sent that were not chosen?

File: solidity/contracts/bridge/Wallets.sol (hunk 1 above)
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/Wallets.sol:583]: Yeah, I was talking about the source wallet and misunderstood you. Regarding target wallets, you're right. However, the purpose of this TODO is just to capture low-level technical details related to this PR and not a full specification of those functions since we have that in the RFC.

File: solidity/contracts/bridge/Wallets.sol (hunk 2 above)
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/Wallets.sol:593]: Nope, they are not in the scope of this PR and will be covered in follow-up PRs. I explained the purpose of those TODOs here: https://github.com/keep-network/tbtc-v2/pull/184#discussion_r842434672

File: solidity/contracts/bridge/Bridge.sol [hunk 3]
```diff
@@ -1189,6 +1205,187 @@ contract Bridge is Ownable, EcdsaWalletOwner {
...
+
+    /// @notice Returns the fraud challenge identified by the given key built
+    ///         as keccak256(walletPublicKey|sighash|v|r|s).
+    function fraudChallenges(uint256 challengeKey)
+        external
+        view
+        returns (Frauds.FraudChallenge memory)
+    {
+        return frauds.challenges[challengeKey];
+    }
```
Comment (tomaszslabon) [inline solidity/contracts/bridge/Bridge.sol:1387]: I've already moved the fraud-related changes in the same spot in https://github.com/keep-network/tbtc-v2/pull/186/commits/1da356d0b4bac0a802c7c1c94157a695d51cb2bd in [another PR](https://github.com/keep-network/tbtc-v2/pull/186).
I think we could just revert these changes in this PR.

File: solidity/contracts/bridge/Bridge.sol [hunk 4]
```diff
@@ -1999,4 +2014,228 @@ contract Bridge is Ownable, EcdsaWalletOwner {
...
+            // <0x1976a914> <20-byte PKH> <0x88ac>.
+            bytes32 targetWalletP2PKHScriptKeccak = keccak256(
+                abi.encodePacked(
+                    hex"1976a914",
+                    targetWalletPubKeyHash,
+                    hex"88ac"
+                )
+            );
+            // Build the expected P2WPKH script which has the following format:
+            // <0x160014> <20-byte PKH>.
```
Comment (tomaszslabon) [inline solidity/contracts/bridge/Bridge.sol:2191]: Could we explain briefly where `<0x160014>` comes from?

File: solidity/contracts/bridge/Bridge.sol (hunk 3 above)
Comment (pdyraga) [reply solidity/contracts/bridge/Bridge.sol]: If #186 gets merged first, I think this change should squash nicely. Same vice-versa.

Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/Bridge.sol]: Already reverted it https://github.com/keep-network/tbtc-v2/pull/184/commits/8ba39e89133b396d7af14e23cc5a7da2586c11b2

File: solidity/contracts/bridge/Bridge.sol (hunk 4 above)
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/Bridge.sol]: Sure! See https://github.com/keep-network/tbtc-v2/pull/184/commits/220f5d1d264c2b90962a34250e7485fe837ed07e

File: solidity/contracts/bridge/Bridge.sol [hunk 5]
```diff
@@ -2005,4 +2020,228 @@ contract Bridge is Ownable, EcdsaWalletOwner {
...
+            // output. To do so, we must extract the full script from the output
+            // and compare with the expected P2PKH and P2WPKH scripts
+            // referring to that 20-byte public key hash. The output consists
+            // of an 8-byte value and a variable length script. To extract the
+            // script we slice the output starting from 9th byte until the end.
+            bytes32 outputScriptKeccak = keccak256(
+                output.slice(8, output.length - 8)
+            );
+            // Build the expected P2PKH script which has the following format:
+            // <0x1976a914> <20-byte PKH> <0x88ac>.
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/Bridge.sol]: Sure! See https://github.com/keep-network/tbtc-v2/pull/184/commits/220f5d1d264c2b90962a34250e7485fe837ed07e

File: solidity/contracts/bridge/Bridge.sol [hunk 6]
```diff
@@ -2010,4 +2040,243 @@ contract Bridge is Ownable, EcdsaWalletOwner {
...
+    ///        See `BitcoinTx.Proof` docs for reference. The `bitcoinHeaders`
+    ///        field must contain a valid number of block headers, not less
+    ///        than the `txProofDifficultyFactor` contract constant.
+    ///      - `mainUtxo` components must point to the recent main UTXO
+    ///        of the given wallet, as currently known on the Ethereum chain.
+    ///        Additionally, the recent main UTXO on Ethereum must be set.
+    ///      - `walletPubKeyHash` must be connected with the main UTXO used
+    ///        as transaction single input.
+    ///      - The wallet that `walletPubKeyHash` points to must be in the
+    ///        Live state.
```
Comment (tomaszslabon) [inline solidity/contracts/bridge/Bridge.sol:2083]: Shouldn't the wallet that `walletPubKeyHash` points be in `MovingFunds` state?

File: solidity/contracts/bridge/Bridge.sol [hunk 7]
```diff
@@ -246,6 +251,11 @@ contract Bridge is Ownable, EcdsaWalletOwner {
     ///         to the redeemer in full amount.
     uint256 public redemptionTimeout;
 
+    /// TODO: Make it governable.
+    /// @notice Maximum amount of the total BTC transaction fee that is
+    ///         acceptable in a single moving funds transaction.
+    uint64 public movingFundsTxMaxTotalFee;
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/Bridge.sol:257]: Nope, there can be multiple outputs. And it is hard to reason about fee per output. This is why I went with the total transaction fee since it makes more sense with moving funds. We do not do any bookkeeping with the received amounts and we just check if the entire transaction has not burned too much amount for fees.

File: solidity/contracts/bridge/Bridge.sol [hunk 8]
```diff
@@ -432,6 +445,9 @@ contract Bridge is Ownable, EcdsaWalletOwner {
         redemptionTreasuryFeeDivisor = 2000; // 1/2000 == 5bps == 0.05% == 0.0005
         redemptionTxMaxFee = 1000; // 1000 satoshi
         redemptionTimeout = 172800; // 48 hours
+        movingFundsTxMaxTotalFee = 10000; // 10000 satoshi
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/Bridge.sol:448]: `redemptionTxMaxFee` is the max fee for the entire transaction. `depositTxMaxFee` is per deposit and `redemptionTxMaxFee` is per redemption. This is why `redemptionTxMaxFee` is much bigger.

File: solidity/contracts/bridge/Bridge.sol [hunk 9]
```diff
@@ -2010,4 +2040,243 @@ contract Bridge is Ownable, EcdsaWalletOwner {
...
+    ///         and to make the necessary state changes. Moving funds is only
+    ///         accepted if it satisfies SPV proof.
+    ///
+    ///         The function validates the moving funds transaction structure
+    ///         by checking if it actually spends the main UTXO of the declared
+    ///         wallet and locks the value on the pre-committed target wallets
+    ///         using a reasonable transaction fee. If all preconditions are
+    ///         met, this functions closes the source wallet.
+    ///
+    ///         It is possible to prove the given moving funds only one time.
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/Bridge.sol]: See https://github.com/keep-network/tbtc-v2/pull/184/commits/d78bef03edcf28d1a51ecfaea5d27a1a2ddf2e77

File: solidity/contracts/bridge/Bridge.sol [hunk 10]
```diff
@@ -1703,37 +1720,35 @@ contract Bridge is Ownable, EcdsaWalletOwner {
         ] = true;
     }
 
-    /// @notice Processes the Bitcoin redemption transaction input vector. It
-    ///         extracts the single input then the transaction hash and output
-    ///         index from its outpoint.
-    /// @param redemptionTxInputVector Bitcoin redemption transaction input
+    /// @notice Parses the input vector of an outbound Bitcoin transaction
+    ///         performed by the given wallet. It extracts the single input then
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/Bridge.sol]: See https://github.com/keep-network/tbtc-v2/pull/184/commits/d78bef03edcf28d1a51ecfaea5d27a1a2ddf2e77

File: solidity/contracts/bridge/Bridge.sol [hunk 11]
```diff
@@ -1703,37 +1720,35 @@ contract Bridge is Ownable, EcdsaWalletOwner {
...
+    ///         the transaction hash and output index from its outpoint.
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/Bridge.sol]: See https://github.com/keep-network/tbtc-v2/pull/184/commits/d78bef03edcf28d1a51ecfaea5d27a1a2ddf2e77

File: solidity/contracts/bridge/Bridge.sol [hunk 12]
```diff
@@ -1645,20 +1661,21 @@ contract Bridge is Ownable, EcdsaWalletOwner {
...
     }
 
-    /// @notice Validates whether the redemption Bitcoin transaction input
-    ///         vector contains a single input referring to the wallet's main
-    ///         UTXO. Reverts in case the validation fails.
-    /// @param redemptionTxInputVector Bitcoin redemption transaction input
+    /// @notice Checks whether an outbound Bitcoin transaction performed by
+    ///         the given wallet has an input vector that contains a single
+    ///         input referring to the wallet's main UTXO. Marks that main UTXO
+    ///         as correctly spent if the validation succeeds. Reverts otherwise.
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/Bridge.sol:1672]: See https://github.com/keep-network/tbtc-v2/pull/184/commits/d78bef03edcf28d1a51ecfaea5d27a1a2ddf2e77

File: solidity/test/data/moving-funds.ts [hunk 13]
```diff
@@ -0,0 +1,587 @@
...
+  chainDifficulty: 21461933,
+}
+
+/**
+ * `SingleTargetWalletButP2SH` test data represents a moving funds with the
+ * following properties:
+ * - 1 main UTXO input
+ * - 1 P2SH output that matches the target wallet from the commitment
+ * - 6+ on-chain confirmations of the sweep transaction
+ */
```
Comment (pdyraga) [inline solidity/test/data/moving-funds.ts:384]: We could note it is not a valid moving funds transaction and that it is going to be rejected by the bridge.

File: solidity/test/data/moving-funds.ts [hunk 14]
```diff
@@ -0,0 +1,587 @@
...
+
+/**
+ * `MultipleTargetWalletsAndDivisibleAmount` test data represents a moving
+ * funds with the following properties:
+ * - 1 main UTXO input
+ * - 2 P2PKH and 1 P2WPKH outputs that matches the target wallets from
+ *   the commitment
+ * - The total transacted amount is distributed unevenly over the target wallets
+ * - 6+ on-chain confirmations of the sweep transaction
+ */
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/test/data/moving-funds.ts:308]: See https://github.com/keep-network/tbtc-v2/pull/184/commits/93fad304d82cbcd8e36f301392dc3e271e623e82

File: solidity/test/data/moving-funds.ts [hunk 15]
```diff
@@ -0,0 +1,587 @@
...
+      "1a10de64d4000060202b62c803c2129fbbf533f35de3e84694ed475aab83dfe11ba0" +
+      "000000000000000de368e3dfdf42f498e8dea274f4c277547844bb07c28340a1a288" +
+      "c86ea38d1113f94e621ec8001a7372433c",
+  },
+
+  chainDifficulty: 21461933,
+}
+
+/**
+ * `MultipleTargetWalletsAndDivisibleAmount` test data represents a moving
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/test/data/moving-funds.ts]: See https://github.com/keep-network/tbtc-v2/pull/184/commits/93fad304d82cbcd8e36f301392dc3e271e623e82

File: solidity/test/data/moving-funds.ts [hunk 16]
```diff
@@ -0,0 +1,587 @@
+import { BigNumberish, BytesLike } from "ethers"
+import { walletState } from "../fixtures"
+
+/**
+ * Represents a set of data used for given moving funds scenario.
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/test/data/moving-funds.ts]: See https://github.com/keep-network/tbtc-v2/pull/184/commits/93fad304d82cbcd8e36f301392dc3e271e623e82

**Review by tomaszslabon — APPROVED** [review]

File: solidity/test/bridge/Bridge.MovingFunds.test.ts [hunk 17]
```diff
@@ -0,0 +1,1005 @@
...
+                              () => {
+                                context(
+                                  "when source wallet is in the MovingFunds state",
+                                  () => {
+                                    context(
+                                      "when target wallets commitment is submitted",
+                                      () => {
+                                        context(
+                                          "when actual target wallets correspond to the commitment",
+                                          () => {
```
Comment (pdyraga) [inline solidity/test/bridge/Bridge.MovingFunds.test.ts:80]: Non-blocking this PR but I think we could un-fold some of a deeply nested structures in unit tests, just like this one.

For example, it does not really matter if we ensure single input points to the wallet's main UTXO if the output vector has P2SH output.

We could first test unhappy paths on the same level of "nesting" and then have a happy path scenario. Something like:
```
when there is no main UTXO for the given wallet
  should revert
when main UTXO data are invalid
  should revert
when input count is other than one
  should revert
when the single input doesn't point to the wallet's main UTXO
  should revert
when the output vector doesn't only reference 20-byte hashes
  should revert
when transaction amount is not distributed evenly
  should revert
when transaction fee is too high
  should revert
when all input data are valid
  when there is a single target wallet
    should mark the main UTXO as correctly spent
    should unset the main UTXO for the source wallet
    should put the source wallet in the Closed state
(...)
```

I know we've been nesting quite heavily in this repo but this example is the most extreme one given it operates on a quite complex validation path.

Anyway, not blocking, we'll have a chance to revisit it when working on target wallets commitment function.

File: solidity/contracts/bridge/Bridge.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/Bridge.sol:257]: Yeah, you are right. I should have my coffee first. Can we make it super-explicit in docs?

I propose adding `@dev` section to `depositTxMaxFee`, `redemptionTxMaxFee`, and `movingFundsTxMaxTotalFee` like:

```
    /// (...)
    /// @dev This is a per-deposit input max fee for the sweep transaction.
    uint64 public depositTxMaxFee;

    /// (...)
    /// @dev This is a per-redemption output max fee for the redemption transaction.
    uint64 public redemptionTxMaxFee;

    /// (...)
    /// @dev This is a TOTAL max fee for the move funds transaction. Note that
    ///      `depositTxMaxFee` is per single deposit and `redemptionTxMaxFee`
    ///      if per single redemption. `movingFundsTxMaxTotalFee` is a total fee
    ///      for the entire transaction.
    uint64 public movingFundsTxMaxTotalFee;
```

File: solidity/contracts/bridge/Bridge.sol (hunk 8 above)
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/Bridge.sol:448]: Right. What I am trying to say is that it *could happen* that we need to sweep a single deposit or that we need to process a single redemption request. Then, the fee per deposit and a fee per redemption will be equal to the fee of an entire Bitcoin transaction. And if we think 1k is not enough given we set `movingFundsTxMaxTotalFee` to 10k, it could happen that 1k will not be enough for `depositTxMaxFee` and `redemptionTxMaxFee`.

[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/Bridge.sol:448]: Yeah but those are just initial values and there is a TODO about revisiting them before the launch. Does it work for now or do you want to change something?

[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/bridge/Bridge.sol:448]: I would change them. I think they should be the same and if we agree on it, it is better to reflect it from the very beginning.

Is it a big change? Does it affect unit tests?

File: solidity/test/bridge/Bridge.MovingFunds.test.ts (hunk 17 above)
Comment (lukasz-zimnoch) [reply solidity/test/bridge/Bridge.MovingFunds.test.ts:80]: I don't think nesting is something bad. We were writing our tests in different ways across Solidity repositories and, in my opinion, nesting works very well because it provides a good specification of how the code is supposed to work. It also allows introducing changes in an easy way. The biggest problem I see with the approach you mentioned is that it doesn't show all the dependencies. For example, in the scenario `when the transaction amount is not distributed evenly` the input vector and a lot of other things must be valid to test this scenario. However, the test structure does not make it clear. Also, we must initialize such preconditions for all cases separately. Now, if we use nesting, you can do initialization once at the appropriate level, and reuse it for dependent cases (a good example is here: https://github.com/keep-network/tbtc-v2/blob/749b9391b64c3a3fcc1eb2a1412a5f5620189e34/solidity/test/bridge/Bridge.Wallets.test.ts#L1075).

File: solidity/contracts/bridge/Bridge.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/Bridge.sol:263]: See https://github.com/keep-network/tbtc-v2/pull/184/commits/8cd1049ee269075e74db2c66626e00216e55a39b

File: solidity/test/bridge/Bridge.MovingFunds.test.ts [hunk 18]
```diff
@@ -0,0 +1,1005 @@
...
+                                "Transaction amount is not distributed evenly"
+                              )
+                            })
+                          }
+                        )
+                      }
+                    )
+
+                    context(
+                      "when the output vector has not only P2PKH and P2WPKH outputs",
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/test/bridge/Bridge.MovingFunds.test.ts]: See https://github.com/keep-network/tbtc-v2/pull/184/commits/855692d82f0c876d1d131fa9d506993d135840c8

File: solidity/test/bridge/Bridge.MovingFunds.test.ts [hunk 19]
```diff
@@ -0,0 +1,1005 @@
...
+                            "Output must be P2PKH or P2WPKH"
+                          )
+                        })
+                      }
+                    )
+                  }
+                )
+
+                context(
+                  "when the output vector doesn't only reference 20-byte hashes",
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/test/bridge/Bridge.MovingFunds.test.ts]: See https://github.com/keep-network/tbtc-v2/pull/184/commits/855692d82f0c876d1d131fa9d506993d135840c8

File: solidity/test/bridge/Bridge.MovingFunds.test.ts (hunk 17 above)
Comment (pdyraga) [reply solidity/test/bridge/Bridge.MovingFunds.test.ts:80]: I am not saying nesting is bad in general. The example you linked is a good use of nesting in my opinion. I just think that in the particular case of `Bridge.MovingFunds.test.ts` we could be nested too deeply, especially given that we are not doing any setup between nested contexts. Just something to think about.

**Review by pdyraga — APPROVED** [review]
